### PR TITLE
Add a new oldest_xmin check.

### PR DESCRIPTION
**Pull request by rjuju**
This implements the check mentioned in https://github.com/OPMDG/check_pgactivity/issues/201.

I'm not extremely sure of the pg_stat_activity part.  The problems I can see are:

- as soon as eg. a transaction retains some xid, active queries started after will also look like they're retaining the same amount of transactions.  However, if t hose queries continue to run after the initial transaction is finished, they'll be the source of retention.  So it's quite hard to really identify what actually retains the xmin horizons from an aggregated view.  The query is already complex enough, so I'm not in favor of adding extra steps to try to be smarter about that.

- For pg_stat_activity, I'm not exactly sure if both backend_xmin *and* backend_xid can both be earlier than the other.  Minimal testing showed than `coalesce(backend_xmin, backend_xid)` was enough, but I certainly didn't try a lot of scenario.

### CODE REVIEW

File: check_pgactivity [hunk 1]
```diff
@@ -5600,6 +5604,167 @@ sub check_oldest_idlexact {
...
+    $w_limit = $args{'warning'};
+
+    @hosts = @{ parse_hosts %args };
+
+    pod2usage(
+        -message => 'FATAL: you must give only one host with service "oldest_xmin".',
+        -exitval => 127
+    ) if @hosts != 1;
+
+    is_compat $hosts[0], 'oldest_xmin', $PG_VERSION_84 or exit 1;
```
Comment (ioguix) [inline check_pgactivity]: I wonder what is the point to support 9.3 and before as we only check for prepared xact. There's already the `oldest_2pc` check that raise an alert for stalled prepared xact.

File: check_pgactivity [hunk 2]
```diff
@@ -5600,6 +5604,167 @@ sub check_oldest_idlexact {
...
+            }
+
+            push @msg_warn => "$r->[0]_age"
+                if ($r->[1] ne 'NaN' and $r->[1] > $w_limit);
+        }
+    }
+
+    return status_critical( $me, [
+        'Critical: '. join(',', @msg_crit)
+        . (scalar @msg_warn? 'Warning: '. join(',', @msg_warn):'')
```
Comment (ioguix) [inline check_pgactivity]: lack a space betwwen the critical list and the `Warning` label?

File: check_pgactivity [hunk 3]
```diff
@@ -154,6 +154,10 @@ my %services = (
         'sub'  => \&check_oldest_idlexact,
         'desc' => 'Check the oldest idle transaction.'
     },
+    'oldest_xmin' => {
+        'sub'  => \&check_oldest_xmin,
+        'desc' => 'Check the xmin horizon per distinct sources of xmin retention.'
```
[the remark this answers is not in the document]
Comment (rjuju) [reply check_pgactivity]: That sounds weird.  **for** distinct...?

File: check_pgactivity [hunk 4]
```diff
@@ -5600,6 +5604,167 @@ sub check_oldest_idlexact {
...
+source doesn't retain any transaction, NaN is returned.
+For versions prior to 9.4, only C<2pc> source of xmin retention is available,
+so other sources won't appear in the perfdata.
+
+Critical and Warning thresholds only accept a raw number of transaction.
+
+Required privileges: an unprivileged role checks only its own queries;
+a pg_read_all_stats (10+) or superuser (<10) role is required to check
+pg_stat_replication.  2PC, pg_stat_activity, and replication slots don't
+require special privileges.
```
[the remark this answers is not in the document]
Comment (rjuju) [reply check_pgactivity]: Do you have a suggestion for a better wording?

File: check_pgactivity [hunk 5]
```diff
@@ -5600,6 +5604,167 @@ sub check_oldest_idlexact {
...
+    my @msg_crit;
+    my @msg_warn;
+    my @hosts;
+    my $c_limit;
+    my $w_limit;
+    my %args        = %{ $_[0] };
+    my $me          = 'POSTGRES_OLDEST_XMIN';
+    my %queries     = (
+        # 8.4 is the first supported version as we rely on window functions to
+        # get the oldest xmin.  Only 2PC has transaction information available
```
[the remark this answers is not in the document]
Comment (rjuju) [reply check_pgactivity]: Probably yes, although I'm not sure it's worth the trouble.

File: check_pgactivity [hunk 6]
```diff
@@ -5600,6 +5604,167 @@ sub check_oldest_idlexact {
...
+            ORDER BY age(xmin) DESC NULLS LAST
+          ) rownum, age(xmin) AS age, xmin
+          FROM raw
+          RIGHT JOIN (
+            SELECT           'query'
+            UNION ALL SELECT 'idle_xact'
+            UNION ALL SELECT 'active_xact'
+            UNION ALL SELECT '2pc'
+            UNION ALL SELECT 'repslot'
+            UNION ALL SELECT 'walsender'
```
[the remark this answers is not in the document]
Comment (rjuju) [reply check_pgactivity]: If you want.  Do you know since when this is supported?

File: check_pgactivity (hunk 1 above)
Comment (rjuju) [reply check_pgactivity]: Because a 1m old 2pc holding 10 transactions is not the same as a 1min old 2pc holding 1M transactions, and if you have a non uniform xid generation it's likely to have random bloat issue that oldest_2pc won't be able to help without a lot of noise?

File: check_pgactivity (hunk 2 above)
Comment (rjuju) [reply check_pgactivity]: Either I'm missing something or the problem exists for every other calls to status_critical.

File: check_pgactivity (hunk 4 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: I'm sure as I did not check what privileges were required from various place. But as far as I understand your wording, what about:

> Required privileges: an unprivileged role can check xmin from 2PC, pg_stat_activity, and 
> replication. a pg_read_all_stats (10+) or superuser (<10) role is required to check
> pg_stat_replication. 

I don't get the difference between `an unprivileged role checks only its own queries`  and `pg_stat_activity don't require special privileges` though...

File: check_pgactivity (hunk 6 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: Since this query is for 9.4 and after, do we care?

[the remark this answers is not in the document]
Comment (rjuju) [reply check_pgactivity]: oh right, indeed.

[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: For the completeness of my answer, this has been introduced in 8.2. See chapter https://www.postgresql.org/docs/8.2/release-8-2.html#AEN80530

File: check_pgactivity (hunk 4 above)
[the remark this answers is not in the document]
Comment (rjuju) [reply check_pgactivity]: Actually after re-reading pg_stat_get_activity() I think that the "an unprivileged role checks only its own queries;" is a missed copy/paste and should be removed, thus this should be ok:

```
Required privileges: a pg_read_all_stats (10+) or superuser (<10) role is required to check
pg_stat_replication.  2PC, pg_stat_activity, and replication slots don't
require special privileges.
```

File: check_pgactivity (hunk 1 above)
Comment (ioguix) [reply check_pgactivity]: Well, ok.

FYI, I can test all versions up to 8.1 where pg_prepared_xacts has been introduced. So you can go the full way down this abyssal version.

File: check_pgactivity (hunk 2 above)
Comment (ioguix) [reply check_pgactivity]: Most of status_critical call I found are using multiple array entries whereas you are doing a string concatenation.

File: check_pgactivity (hunk 5 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: It will be the same query for all version down to 8.1:

~~~
postgres@postgres=# SELECT version();
 PostgreSQL 8.1.23 on x86_64-unknown-linux-gnu, compiled by GCC gcc (Debian 8.3.0-6) 8.3.0

postgres@postgres=# \d pg_prepared_xacts 
 transaction | xid                      | 
 gid         | text                     | 
 prepared    | timestamp with time zone | 
 owner       | name                     | 
 database    | name                     | 
~~~

File: check_pgactivity (hunk 2 above)
Comment (rjuju) [reply check_pgactivity]: I see.  Somehow I managed to copy/paste the only broken call: https://github.com/OPMDG/check_pgactivity/blob/master/check_pgactivity#L5170-L5174

File: check_pgactivity (hunk 1 above)
Comment (rjuju) [reply check_pgactivity]: Thanks!

File: check_pgactivity (hunk 3 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: as discussed, I still prefer __from__

File: check_pgactivity (hunk 4 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: works for me.

File: check_pgactivity (hunk 2 above)
Comment (ioguix) [reply check_pgactivity]: Oops.

Fixed: a09b8b823bd0529

Thanks for the report :)

**Comment by rjuju** [comment]
As discussed off thread, I realized that I forgot to also handle the impacted databases.  I just pushed an additional commit (that should be squashed, after a rebase I'll take care of that once everything is ok) to fix both that and @ioguix comments.

File: check_pgactivity (hunk 3 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity]: This still need to be fixed.

File: check_pgactivity [hunk 7]
```diff
@@ -5600,6 +5604,186 @@ sub check_oldest_idlexact {
 }
 
 
+=item B<oldest_xmin> (8.4+)
+
+Check the xmin I<horizon> from distinct sources of xmin retention.
+
+Perfdata contains the oldest xmin and maximum age for the following source of
+xmin retention: C<query> (a currently executing query), C<active_xact> (an
+opened transaction currently executin a query), C<idle_xact> (an opened
```
Comment (ioguix) [inline check_pgactivity]: > currently executin**g**

File: check_pgactivity [hunk 8]
```diff
@@ -5600,6 +5604,186 @@ sub check_oldest_idlexact {
...
+transaction being idle), C<2pc> (a pending prepared transaction), C<repslot> (a
+recpliation slot) and C<walwender> (a WAL sender replication process), for each
```
Comment (ioguix) [inline check_pgactivity]: > C<wal**s**ender>

File: check_pgactivity [hunk 9]
```diff
@@ -5600,6 +5604,186 @@ sub check_oldest_idlexact {
...
+          ) f(kind) ON raw.kind = f.kind
+          -- we use this JOIN condition to make sure that we'll always have a
+          -- full record for all (connectable) databases
+          JOIN pg_database d ON d.datname = coalesce(raw.datname, d.datname)
+          WHERE d.datallowconn
+        )
+        SELECT kind, datname, age, xmin
+        FROM ordered
+        WHERE rownum = 1
+        }
```
Comment (ioguix) [inline check_pgactivity]: This query reports as many walsender than existing database, which is wrong.

File: check_pgactivity (hunk 7 above)
Comment (rjuju) [reply check_pgactivity]: will fix

File: check_pgactivity (hunk 8 above)
Comment (rjuju) [reply check_pgactivity]: will fix

File: check_pgactivity (hunk 9 above)
Comment (rjuju) [reply check_pgactivity]: This returns as many physical walsender as connectable database, which is exactly what is expected.  It simplifies knowing what is the xmin horizon on a given database without checking the possible "global" limits.

Comment (ioguix) [reply check_pgactivity]: physical walsender are global, they are not related to one single database. So I don't understand your answer. It's still confusing

File: check_pgactivity (hunk 3 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity:159]: whatever, this is not very important.

File: check_pgactivity (hunk 5 above)
[the remark this answers is not in the document]
Comment (ioguix) [reply check_pgactivity:5641]: well, ok, I give up on that as well.

File: check_pgactivity [hunk 10]
```diff
@@ -5600,6 +5604,186 @@ sub check_oldest_idlexact {
...
+
+    is_compat $hosts[0], 'oldest_xmin', $PG_VERSION_84 or exit 1;
+
+    @rs = @{ query_ver( $hosts[0], %queries ) };
+
+    REC_LOOP: foreach my $r (@rs) {
+        map { $_ = 'NaN' if $_ eq ''} @{$r}[2..3];
+        push @perfdata => (
+            ["$r->[0]_$r->[1]_age", $r->[2]],
+            ["$r->[0]_$r->[1]_xmin", $r->[3]]
```
Comment (ioguix) [inline check_pgactivity:5761]: I still think reporting 6 kinds * 2 metrics * num_db  is very a lot of metrics, but anyway. I can give up on this as far as we can move on this PR. We'll be able to discuss and find solutions later if it is a real pb raising from the field.

However, using `"$r->[1]_$r->[0]_age` (`$db_$kind_age|xmin`) reads better to me. Eg. `dbname_walsender_age` instead of `walsender_dbname_age`. "age" is related to "kind". Moreover, it is easier to sort by dbname when checking metrics by hands.

Comment (rjuju) [reply check_pgactivity:5767]: The only way to reduce then number of metrics is to report physical slots only once, as this is the only global metric.  That probably won't save much, but I won't object to that change.  Special care will have to be taken for the documentation to help users determining what is the xmin horizon on a specific database, as this won't be as obvious as before.

File: check_pgactivity (hunk 9 above)
Comment (ioguix) [reply check_pgactivity:5837]: I understood why it has been done this way after discussing. Closing this conversation

File: check_pgactivity (hunk 10 above)
Comment (ioguix) [reply check_pgactivity:5876]: As discussed, I'm ok with this.

**Review by ioguix — APPROVED** [review]

**Comment by ioguix** [comment]
Hey @rjuju,

I've been reviewing and testing. I closed all our active conversations as we finished discussing all of them.

I fixed a bug in the query were some `NaN` perfdata (`idle_xact` or `active_xact` or `query`) were not shown for database when they were set in other databases.

Please, do some test/review and fixup/merge when you are ok.

**Comment by ioguix** [comment]
BTW, with only 4 connectable db (including `template1` and `postgres`), I gathered 48 perfdatas. 12 metrics per database.

Maybe we can add a `--summary` arg or equivalent to only provide in perfdata the oldest xmin/age per database, and report in long messages what is the oldest horizon for each database?

**Comment by frost242** [comment]
+1 for a `--summary` option.

**Comment by rjuju** [comment]
Thanks @ioguix 
I'm fine with a `--summary` option.  However, just to be clear I don't expect anyone using that check for all databases nor for a long time.  It's really a diagnosis check to help track down bloat issue happening at some random interval, but the rest of the time you don't really need to monitor it.

I'll review the changes soon.

**Comment by ioguix** [comment]
As discussed on IRC, I added a few commits:

* thresholds are now optionnal
* the check is now terse by default: one perfdata (oldest age) per database plus one info per database (what is the kind of the oldest xmin) in messages if there's some data to show.
* new arg `--detailed` throw detailed perfdatas (12) for each database

Please, review.

++

**Comment by rjuju** [comment]
I'm fine with the new behavior.  About the patch, I think we should document the new `--detailed` option along with the other ones. 

I also see that you don't check for compatibility with this option.  Is that intended?  I'm not sure if anyone will blindly add a `--detailed` everywhere just in case so that's may not be a huge problem, but until now our current approach is to explicitly check for every single option compatibility.

**Comment by ioguix** [comment]
>  About the patch, I think we should document the new --detailed option along with the other ones.

Well, a lots of specifics argument are not documented with other, normal, ones. Specifics ones are only documented in their related service. Eg.: `--query`, `--path`, `--type`, `--suffix`, `--uid`, GUCs, `--unarchiver`, etc.

> I also see that you don't check for compatibility with this option. Is that intended?

Fixed in last commit.
